broker-daemon/block-order-worker/index.js

```diff
@@ -316,6 +281,57 @@ class BlockOrderWorker extends EventEmitter {
     }
   }
 
+  /**
+   * Move a block order to a completed state if all orders have been completed
+   * @todo What should we do if this method fails, but the order itself is completed?
+   * @param {String} blockOrderId
+   */
+  async completeBlockOrder (blockOrderId) {
+    this.logger.info('Attempting to put block order in a completed state', { id: blockOrderId })
+
+    const blockOrder = await this.getBlockOrder(blockOrderId)
+
+    // check the fillamount on each collection of state machines from the block order
+    // and make sure that either is equal to how much we are trying to fill.
+    let totalFilled = Big(0)
+
+    if (blockOrder.fills.length) {
+      totalFilled = blockOrder.fills.reduce((acc, fsm) => acc.plus(fsm.fill.fillAmount), Big(0))
+    } else if (blockOrder.openOrders.length) {
+      totalFilled = blockOrder.openOrders.reduce((acc, osm) => {
+        // If the order has not been filled yet, then the `fillAmount` will be undefined
+        // so we instead default to 0
+        return acc.plus(osm.order.fillAmount || 0)
+      }, Big(0))
+    } else {
+      this.logger.error('Broker in a bad state, no fills or orders on block order', { blockOrderId })
+      throw new Error('No fills or orders on blockOrder')
+    }
+
+    this.logger.debug('Current total filled amount: ', { totalFilled, blockOrderAmount: blockOrder.baseAmount })
+
+    const stillBeingFilled = totalFilled.lt(blockOrder.baseAmount)
+
+    const activeOrders = blockOrder.openOrders.some(osm => osm.order.active)
+    const activeFills = blockOrder.fills.some(fsm => fsm.fill.active)
+
+    if (!activeOrders && !activeFills && !stillBeingFilled) {
+      // An order is only completed if all orders underneath the blockorder are out of
+      // an `ACTIVE` state and it is entirely filled
+      blockOrder.complete()
+      await promisify(this.store.put)(blockOrder.key, blockOrder.value)
+      this.logger.info('Moved block order to completed state', { blockOrderId })
+    } else if (stillBeingFilled && !activeFills && !activeOrders) {
+      this.logger.error('Block order is still being filled, but has no active orders/fills', { blockOrderId })
+      throw new Error('Block order in a weird state. Order is filled, but no active orders/fills still exist')
```

Comment:
@treygriffith thats why we throw the error. If its in a weird state, we want to know and need to fix the bug. I dont think we should avoid this state because of a race-condition. We can simply re-run this method and the order will effectively complete, but it is still nice to know if this situation occurs.

Either way, if I should remove the check, that is fine too.

Reply:
@treygriffith I've had a second thought. If this piece of code makes the `complete` method too complicated (or weird) then we should remove it. If this issue occurs, it is probably better to realize and error at a different point in the lifecycle, than at the very end. What do you think?